### PR TITLE
haxe: 3.4.6 -> 4.2.1

### PR DESCRIPTION
**Pull request by locallycompact**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/extlib/default.nix [hunk 1]
```diff
@@ -3,11 +3,11 @@
 assert lib.versionAtLeast (lib.getVersion ocaml) "3.11";
 
 stdenv.mkDerivation {
-  name = "ocaml${ocaml.version}-extlib-1.7.7";
+  name = "ocaml${ocaml.version}-extlib-1.7.8";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/extlib/default.nix:6]:
```suggestion
  pname = "ocaml${ocaml.version}-extlib";
  version = "1.7.8";
```

File: pkgs/development/ocaml-modules/extlib/default.nix [hunk 2]
```diff
@@ -3,11 +3,11 @@
 assert lib.versionAtLeast (lib.getVersion ocaml) "3.11";
 
 stdenv.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/extlib/default.nix:5]:
```suggestion
stdenv.mkDerivation rec {
```

File: pkgs/development/ocaml-modules/extlib/default.nix [hunk 3]
```diff
@@ -3,11 +3,11 @@
...
-  name = "ocaml${ocaml.version}-extlib-1.7.7";
+  name = "ocaml${ocaml.version}-extlib-1.7.8";
 
-  src = fetchurl {
-    url = "http://ygrek.org.ua/p/release/ocaml-extlib/extlib-1.7.7.tar.gz";
-    sha256 = "1sxmzc1mx3kg62j8kbk0dxkx8mkf1rn70h542cjzrziflznap0s1";
+ src = fetchurl {
+    url = "https://ygrek.org/p/release/ocaml-extlib/extlib-1.7.8.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/extlib/default.nix:9]:
```suggestion
    url = "https://ygrek.org/p/release/ocaml-extlib/extlib-${version}.tar.gz";
```

File: pkgs/development/ocaml-modules/luv/default.nix [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{ lib, buildDunePackage, fetchurl, ctypes }:
+
+buildDunePackage rec {
+  pname = "luv";
+  version = "0.5.6";
+  useDune2 = true;
+
+  src = fetchurl {
+    url = "https://github.com/aantron/luv/releases/download/0.5.6/luv-0.5.6.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/luv/default.nix:9]:
```suggestion
    url = "https://github.com/aantron/luv/releases/download/${version}/luv-${version}.tar.gz";
```

File: pkgs/development/ocaml-modules/luv/default.nix [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    sha256 = "119nv250fsadvrs94nwzk7qvlwr0kvcpkbwcmfkh13byg8nhkn1m";
+  };
+
+  propagatedBuildInputs = [ ctypes ];
+
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/luv/default.nix:15]:
```suggestion
  meta = with lib; {
```

File: pkgs/development/ocaml-modules/luv/default.nix [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    homepage = "https://github.com/aantron/luv";
+    description = "Binding to libuv: cross-platform asynchronous I/O";
+    license = lib.licenses.lgpl21;
+    maintainers = [ lib.maintainers.vbgl ];
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/luv/default.nix:19]:
```suggestion
    license = licenses.lgpl21;
    maintainers = [ maintainers.vbgl ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by kvtb** [comment]
neko 2.2.0 -> 2.3.0 is required

**Comment by kvtb** [comment]
https://github.com/HaxeFoundation/haxe#version-compatibility

**Comment by sternenseemann** [comment]
@ofborg build google-drive-ocamlfuse

**Comment by sternenseemann** [comment]
* Please split up your updates into individual commits, i. e. the `extlib` update shouldn't be squashed together with the haxe update
* The extlib breaks quite a few things, see #110374 for a discussion of the problems. Not sure if the situation has changed and we can update extlib without breaking stuff.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/luv/default.nix [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+buildDunePackage rec {
+  pname = "luv";
+  version = "0.5.6";
+  useDune2 = true;
+
+  src = fetchurl {
+    url = "https://github.com/aantron/luv/releases/download/${version}/luv-${version}.tar.gz";
+    sha256 = "119nv250fsadvrs94nwzk7qvlwr0kvcpkbwcmfkh13byg8nhkn1m";
+  };
+
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/luv/default.nix:12]: Allow configure script to use `file`

```suggestion

  postConfigure = ''
    for f in src/c/vendor/configure/{ltmain.sh,configure}; do
      substituteInPlace "$f" --replace /usr/bin/file file
    done
  '';

  nativeBuildInputs = [ file ];
  
```

File: pkgs/top-level/ocaml-packages.nix [hunk 8]
```diff
@@ -572,6 +572,8 @@ let
 
     lua-ml = callPackage ../development/ocaml-modules/lua-ml { };
 
+    luv = callPackage ../development/ocaml-modules/luv { };
```
Comment (sternenseemann) [inline pkgs/top-level/ocaml-packages.nix:575]: Make sure `file` comes from top-level, it's easily imaginable we'll have a ocaml package named `file` one day.

```suggestion
    luv = callPackage ../development/ocaml-modules/luv {
      inherit (pkgs) file;
    };
```

**Review by sternenseemann — CHANGES_REQUESTED** [review]

**Comment by locallycompact** [comment]
> * Please split up your updates into individual commits, i. e. the `extlib` update shouldn't be squashed together with the haxe update
> * The extlib breaks quite a few things, see #110374 for a discussion of the problems. Not sure if the situation has changed and we can update extlib without breaking stuff.

This was the last thing that was said in that issue so that's what I did.

https://github.com/NixOS/nixpkgs/pull/110374#issuecomment-777062907

**Comment by sternenseemann** [comment]
Other packages are broken by extlib 1.7.8 we also need to care about. See the linked PR for a discussion of possible solutions. It seems we may need to keep to versions around after all if haxe 4.2.1 can't be made to work with a (patched) version of extlib 1.7.7.

File: pkgs/development/compilers/neko/default.nix [hunk 9]
```diff
@@ -32,7 +32,7 @@ stdenv.mkDerivation rec {
     description = "A high-level dynamically typed programming language";
     homepage = "https://nekovm.org";
     license = licenses.lgpl21;
-    maintainers = [ maintainers.marcweber ];
+    maintainers = [ maintainers.locallycompact ];
```
[the remark this answers is not in the document]
Comment (locallycompact) [reply pkgs/development/compilers/neko/default.nix:35]: He seems unresponsive and I have been wondering whether there would be a haxe update for several years already.

**Comment by locallycompact** [comment]
> Other packages are broken by extlib 1.7.8 we also need to care about. See the linked PR for a discussion of possible solutions. It seems we may need to keep to versions around after all if haxe 4.2.1 can't be made to work with a (patched) version of extlib 1.7.7.

Would you trigger those packages so I can see?

File: pkgs/development/compilers/neko/default.nix (hunk 9 above)
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/compilers/neko/default.nix:35]: Leave them in for now, maintainer activity is not policed in nixpkgs afaik, they can always remove themselves.

File: pkgs/development/ocaml-modules/luv/default.nix [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{ lib, buildDunePackage, fetchurl, ctypes }:
+
+buildDunePackage rec {
+  pname = "luv";
+  version = "0.5.6";
+  useDune2 = true;
+
+  src = fetchurl {
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/luv/default.nix:8]: Never mind, I tried this out locally and tests take at least 12h or so if they ever terminate o_O

**Comment by sternenseemann** [comment]
> Would you trigger those packages so I can see?

In nixpkgs only google-drive-ocamlfuse, it seems.

My suggestion would be to add extlib 1.7.8 as a separate attribute for now and leave it as is until upstream sorts out their next release which should include a proper compatibility layer for the stuff that is currently broken (see <https://github.com/ocaml/opam-repository/pull/17999>).

File: pkgs/development/compilers/neko/default.nix [hunk 11]
```diff
@@ -4,11 +4,11 @@
 
 stdenv.mkDerivation rec {
   pname = "neko";
-  version = "2.2.0";
+  version = "v2-3-0";
 
   src = fetchurl {
-    url = "https://nekovm.org/media/neko-${version}-src.tar.gz";
-    sha256 = "1qv47zaa0vzhjlq5wb71627n7dbsxpc1gqpg0hsngjxnbnh1q46g";
+    url = "https://github.com/HaxeFoundation/neko/archive/${version}.tar.gz";
```
[the remark this answers is not in the document]
Comment (zseri) [reply pkgs/development/compilers/neko/default.nix:10]: wouldn't it be more appropriate to use `fetchFromGitHub` for that?

[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/compilers/neko/default.nix:10]: Agreed, `fetchurl` is not appropriate here, good catch.

File: pkgs/development/ocaml-modules/luv/default.nix (hunk 7 above)
Comment (zseri) [reply pkgs/development/ocaml-modules/luv/default.nix:12]: shouldn't `file` be `${file}/bin/file` instead?

**Comment by sternenseemann** [comment]
It’s in PATH, so not necessary.

File: pkgs/development/ocaml-modules/luv/default.nix (hunk 7 above)
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/luv/default.nix:12]: It’s in `PATH`, so not necessary.

**Comment by sternenseemann** [comment]
@locallycompact I've dabbled around with your PR today and managed to redo it so there's no issue with extlib and we don't need to remove the old compilers just yet: #116953 Mind testing / reviewing?

**Comment by locallycompact** [comment]
@sternenseemann Not at all, thank you very much. I was about to. :)

File: pkgs/development/compilers/neko/default.nix [hunk 12]
```diff
@@ -4,11 +4,11 @@
...
   pname = "neko";
-  version = "2.2.0";
+  version = "2.3.0";
 
   src = fetchurl {
-    url = "https://nekovm.org/media/neko-${version}-src.tar.gz";
-    sha256 = "1qv47zaa0vzhjlq5wb71627n7dbsxpc1gqpg0hsngjxnbnh1q46g";
+    url = "https://github.com/HaxeFoundation/neko/archive/v${builtins.replaceStrings [ "." ] [ "-" ] version}.tar.gz";
+    sha256 = "sha256-hQ5+MXva8k7WUu/v+JwcshOAyhnyDmiilshPa61O6ZU=";
   };
```
Comment (zseri) [inline pkgs/development/compilers/neko/default.nix:12]: suggestion:
```nix
src = fetchFromGitHub {
  owner = "HaxeFoundation";
  repo = "neko";
  rev = "v" + (builtins.replaceStrings ["."] ["-"] version);
  # the following line might be incorrect, needs to be verified
  sha256 = "hQ5+MXva8k7WUu/v+JwcshOAyhnyDmiilshPa61O6ZU=";
};
```

**Comment by sternenseemann** [comment]
Superseded by #116953.

**Pull request closed without merging**